mgmtd: Add new FRR Management Framework and Daemon

pushpasis opened the pull request:
This PR brings in the new YANG-based FRR Management Framework that has following sub-components in it:
 
- A new Management Daemon (from now on referred to as 'mgmtd' or 'MGMTD') that will ultimately become only entity to handle all data operations requests on FRR Management data.
- A new MGMT Frontend Interface for all Frontend/UI clients to send their management data related requests to send to.
- A new MGMT Backend Interface for all backend application daemons/components to connect to and handle/present their application-specific configurational and operational data.
 
NOTE: Currently this PR only introduces the necessary functionalities for handling configuration data. Necessary functionalities for handling/retrieving operational data shall be added in a future PR.
 
This PR also introduces Staticd as a backend client for the MGMTd framework. All the static configuration commands will be diverted to the new MGMTd daemon and will use the Transactional model to make changes to the internal state. Similar mechanism can be used by other daemon to use the MGMT framework in the future.
 
This PR has been split into following commits to facilitate review.
mgmtd: Bringup new MGMTD daemon and add support for Database Module.
- mgmtd: Add MGMT Frontend Interface Framework
- mgmtd: Add MGMT Backend Interface Framework
- mgmtd: Add MGMT Transaction framework
- mgmtd: Add Staticd as the first backend client for the MGMT Framework
- tests: topotests for mgmtd.
- doc: Add documentation for MGMT daemon
 
Co-authored-by: Pushpasis Sarkar <pushpasis@gmail.com>
Co-authored-by: Abhinay Ramesh <rabhinay@vmware.com>
 Co-authored-by: Ujwal P <ujwalp@vmware.com>
 Co-authored-by: Naveen G <nguggarigoud@vmware.com>

Signed-off-by: Yash Ranjan <ranjany@vmware.com >

pguibert6WIND commented:
pull request 10000. congrats :+1:

qlyoung commented:
@pushpasis do you plan to fix the basic issues with the code, such as the use of banned functions, missing signed-off-by line, incorrect commit message etc., in this PR?

pushpasis commented:
> @pushpasis do you plan to fix the basic issues with the code, such as the use of banned functions, missing signed-off-by line, incorrect commit message etc., in this PR?

Hi @qlyoung. Yes we are currently working on it. Need a week's time to take care of all. By the way we need to have access to a arm7 build machine to verify the fixes for the build failures on 32-bit arm platforms. Would you or anyone have one handy?

Thanks
-Pushpasis

donaldsharp commented:
1) #include <zebra.h> should be the first include in pretty much all .c files
2) I see inconsistencies between library header files being included.  Please standardize to our normal standard.  Headers in lib should be `#include "thread.h"`.  Headers in mgmtd should be `#include "mgmtd/mgmtd.h"`
3) On a very casual inspection I saw repeated defines/functions that are available in lib.  MAX and monotime are on the top of what I saw.  We should just use the lib versions.  There is no need to do this 2 times
4) No typedefs
5) I see defines for loop variables added.  Sometimes we have spaces before the opening paranthesis and sometimes we don't.  Please add to the .clang-format these new additions and be consistent with the rest of the code base
6) Open paranthesis are on the same line as the loop.
7) I see enum's being switched with default case statements .  Remove the default and use the compiler to ensure that all cases are being covered.
8) I see blanket includes of header files in lib.  That I would be willing to bet those headers are not being used.  This should be cleaned up
9) Please remove all #if 0 or #if 1.  Dead code is dead.  If needed for the future then provide a #define that is meaningful to the next person behind you

I've just done a very casual inspection of the code and will look over it again, once updates have been made.  I also have an expectation that each step self compiles.  Git bisect ability is very important to be able to do.

pushpasis commented:
> 1. #include <zebra.h> should be the first include in pretty much all .c files
> 2. I see inconsistencies between library header files being included.  Please standardize to our normal standard.  Headers in lib should be `#include "thread.h"`.  Headers in mgmtd should be `#include "mgmtd/mgmtd.h"`
> 3. On a very casual inspection I saw repeated defines/functions that are available in lib.  MAX and monotime are on the top of what I saw.  We should just use the lib versions.  There is no need to do this 2 times
> 4. No typedefs
> 5. I see defines for loop variables added.  Sometimes we have spaces before the opening paranthesis and sometimes we don't.  Please add to the .clang-format these new additions and be consistent with the rest of the code base
> 6. Open paranthesis are on the same line as the loop.
> 7. I see enum's being switched with default case statements .  Remove the default and use the compiler to ensure that all cases are being covered.
> 8. I see blanket includes of header files in lib.  That I would be willing to bet those headers are not being used.  This should be cleaned up
> 9. Please remove all #if 0 or #if 1.  Dead code is dead.  If needed for the future then provide a #define that is meaningful to the next person behind you
> 
> I've just done a very casual inspection of the code and will look over it again, once updates have been made. I also have an expectation that each step self compiles. Git bisect ability is very important to be able to do.

Hi @donaldsharp, thanks a lot for the comments. We will definitely take care of all these and fix it in the next update. As requested by you we are working on fixing the original commits that introduced these issues.

Thanks
-Pushpasis

idryzhov commented:
The first big question I have is why do we need protobuf here.

Looking at the proto file and the code, I don't see any advantages. You don't even use features that protobuf provides, for example, in `BckndMessage` and `FrntndMessage` you have `oneof message` and `Type type`. This is exactly what we would do if we were using C structures. But that's not needed in protobuf, because `oneof` already provides you with this information in `_case` field.

And I see the following disadvantages:
- additional library requirement,
- code style – protobuf-c naming conventions are very different from what we have in FRR,
- complexity for development – I can't simply look at structure definitions, I need to compile the proto file first,
- additional packing/unpacking -> more memory, more CPU cycles.

pushpasis commented:
> The first big question I have is why do we need protobuf here.
> 
> Looking at the proto file and the code, I don't see any advantages. You don't even use features that protobuf provides, for example, in `BckndMessage` and `FrntndMessage` you have `oneof message` and `Type type`. This is exactly what we would do if we were using C structures. But that's not needed in protobuf, because `oneof` already provides you with this information in `_case` field.
> 
> And I see the following disadvantages:
> 
> * additional library requirement,
> * code style – protobuf-c naming conventions are very different from what we have in FRR,
> * complexity for development – I can't simply look at structure definitions, I need to compile the proto file first,
> * additional packing/unpacking -> more memory, more CPU cycles.
> 
> We already have everything we need for inter-daemon communication – look at zapi/zclient. I don't see any reason to reinvent the wheel.

Hi Igor,

I think we discussed this already in the YNG meeting. But off course we can discuss this once again in the next YANG meeting. Protobuf was used due to following two major reasons.

1. We don't need to write complex encode/decode functions. So protobuf helps with the rapid development. Also it takes care of things like 'variable sized arrays' in the messages (which are not going to be easy if we write the decoding/encoding logic ourselves). Also message packaging efficiency is supposed to better in protobuf than using plain C structures.
2. The MGMTd is exposing a message-based API interface towards the Front-end and Back-end clients. If we use C-based structures for message definitions only, we will have to enforce only C-based applications as Frontend (and Backend) clients. While Backend clients will mostly be C-based FRR daemons, IMO it will be not be correct to assume all Front-end clients are going to be entirely implemented in C. To make it language independent, we need to define each message type explicitly. There are two ways to do it, in my mind: a) Define them like rfc style, or b) Define it as protobuf specification. We chose protobuf, since it includes advantages mentioned in 1. Also, the protobuf encode/decode code can now be generated for other languages to facilitate any client implementation in other languages that is supported by protobuf. 

With regards to your comment on 'Type' field, I will take it and remove it from the protobuf specification. Let's discuss more on the next YANG meeting.

Thanks
-Pushpasis

idryzhov commented:
1. This is a single-time job. There are not many messages to encode/decode, and they are not complex. What problem do you see in variable sized arrays? You send the number of elements and then send the elements themselves. You can look at `zapi_route_encode` – it sends the nexthops this way.
2. For backends, we're much more tied to C because of libzebra and zapi. For frontends, yes, it may be convenient to have a protobuf API, but still, there are not too many messages we need to support and they are pretty simple. Every language is capable of processing byte stream received from socket, so we're not making it language-dependent if we're not using protobuf.

I personally hate how protobuf is implemented in C, so I am biased in this discussion. If the community is fine with hard protobuf dependency and fine with the code and style – we'll go with protobuf. Let's discuss it on Monday call.

mwinter-osr reviewed (changes requested):
Documentation update needed. The building-*.rst files need to be updated to include new requirements.

pushpasis commented:
> Documentation update needed. The building-*.rst files need to be updated to include new requirements.

Hi @mwinter-osr we have updated the additional package installation requirements in all the building manuals except for openwrt. We have updated the extra packages needed for docker images too under docker/ folder. Please take a look and let us know if anything is missing.

mwinter-osr commented:
CI:rerun

mwinter-osr reviewed (commented):
CI is updated, but build fails on multiple platforms.

pushpasis commented:
ci: rerun

pushpasis commented:
CI:rerun

mwinter-osr commented on Makefile.am:
```diff
@@ -244,6 +265,7 @@ EXTRA_DIST += \
 	snapcraft/helpers \
 	snapcraft/snap \
 	babeld/Makefile \
+	mgmtd/Makefile
```
Missing the \ at the end of the line.
This causes to the 'make dist' to generate incomplete files and breaks the package build on the RPM based systems on the CI

pushpasis replied:
@mwinter-osr Thanks a lot for pointing that out. Let me fix it asap. Assuming centos8 and fedora29 build will be successful after that, we still need to discuss the centos7 case.

pushpasis replied:
Ok. I have pushed the fix. Hopefully the centos8 and fedora29 builds will not fail again.

riw777 commented on lib/northbound.c:
```diff
@@ -538,10 +543,16 @@ static inline void nb_config_diff_dnode_log(const char *context,
+ * the DB being compared against 'config1'. Typically 'config1'
+ * should be the Running DB and 'config2' is the Candidate DB.
+ */
+void nb_config_diff(const struct nb_config *config1,
```
given the comment above, it might make more sense to rename these to something other than `config1` and `config2` ... maybe `config_ref` and `config_compare` or something similar?

pushpasis replied:
Sure will rename these as per your suggestion. @ranjanyash54 please take care of addressing this comment.

riw777 reviewed (commented):
First pass through ... code looks good. I don't know the vtysh and build parts, so I didn't really spend a lot of time on those. One really minor nit. I'll take another pass over this in the next couple of weeks.

pushpasis commented:
ci: rerun

pushpasis commented:
ci: rerun

pushpasis commented:
CI: rerun

pushpasis commented:
CI: rerun

pushpasis commented:
CI: rerun

ranjanyash54 commented:
ci: rerun

pushpasis commented:
CI: rerun

ranjanyash54 commented:
ci: rerun

pushpasis commented:
@riw777 @idryzhov @mwinter-osr Request you to review the code changes. The CI is successful now.

choppsv1 commented on Makefile.am:
```diff
@@ -244,6 +265,7 @@ EXTRA_DIST += \
 	snapcraft/helpers \
 	snapcraft/snap \
 	babeld/Makefile \
+	mgmtd/Makefile \
```
it isn't perfect sorting, but the daemons are sorted, I suggest putting this between `lib/Makefile` and `nhrpd/Makefile`

pushpasis replied:
Sure will take care of in the next force-push.

pushpasis commented:
ci:rerun rule out intermittent failures

pushpasis commented:
ci:rerun build failed due to ci infra issue

eqvinox commented on mgmtd/mgmt_db.c:
```diff
@@ -0,0 +1,1040 @@
+};
+
+struct mgmt_cmt_info_t {
+	struct mgmt_cmt_info_dlist_item cmt_dlist;
```
What exactly is this, why is it called a database, and why is it not using something like SQLite, GDBM or LMDB?

pushpasis replied:
This is a linked list of commit buffers (upto 10 maximum, with oldest entry removed after 10). Entries in each entry of this list points to a file on disk that has a YANG/JSON based snapshot of the running DB after a specific commit. If not clear already we are supporting commit rollback functionality with capability to rollback upto 10 last commits. Hope it answers your question.

pushpasis replied:
Sure we will not call the commit history buffer any kind of database.

The commit buffers are written out to the a file at the end of each commit. Also the current state of running DB is written out to the disk at the end of commit. There's a very small window between applying the configs and writing it out to the disk. If power goes out in between, the router will restart back from the older config already present on the disk. The recent commit will be lost. Maybe in al later PR we can come up with a solution for that too. But at the moment don't want to complicate the implementation further.

eqvinox commented on mgmtd/mgmt_db.h:
```diff
@@ -0,0 +1,432 @@
+
+struct mgmt_db_ctxt;
+
+PREDECL_DLIST(mgmt_cmt_info_dlist);
```
The type of container should not be encoded into the name, i.e. there should not be a `_dlist` here. Same issue in multiple other places. Please remove all such suffixes.

(Rationale: the containers are explicitly designed to help in changing to another container type. If the type is encoded into the name, this breaks.)

pushpasis replied:
Sure, we will remove _dlist. Can we remove it to mgmt_cmt_info_list? That seems to me is more generic.

pushpasis replied:
I mean using _list should be okay. But not _slist or _dlist.

eqvinox replied:
`_list` is counterintuitive too — if it's e.g. changed into a RB-tree later, it will be very confusing.

I would probably pluralize the names with an `s` at the end, e.g. `mgmt_cmt_infos`.

pushpasis replied:
Ok. Sounds okay to me. Will take care of these in upcoming force push.

eqvinox commented on mgmtd/mgmt_db.h:
```diff
@@ -0,0 +1,432 @@
+
+#define MGMTD_MAX_COMMIT_FILE_PATH_LEN 55
+#define MGMTD_COMMIT_FILE_PATH "/etc/frr/commit-%s.json"
+#define MGMTD_COMMIT_INDEX_FILE_NAME "/etc/frr/commit-index.dat"
```
`/etc` is not a path to create database-type files under.

pushpasis replied:
What would you suggest then? Whatever you suggest the path should be writable from FRR daemons. I think we couple of other paths but they were not writable from mgmtd.

eqvinox replied:
`/var/lib/frr/<namespace>/` is the "correct" place for this according to [FHS](https://en.wikipedia.org/wiki/Filesystem_Hierarchy_Standard)

> State information. Persistent data modified by programs as they run (e.g., databases, packaging system metadata, etc.).

eqvinox replied:
Also this must respect `--localstatedir` from `./configure`

pushpasis replied:
Hmm. Good point let us see how we can make it happen.

pushpasis commented:
> ## 🛑 Basic BGPD CI results: FAILURE
> Results table
> _	_
> Result	FAILURE git merge/10000 [65d8a28](https://github.com/FRRouting/frr/commit/65d8a282e44b913326e83ca1f6c45fd952eebc25) frr.github Build
> Date	03/29/2022
> Start	01:41:25
> Finish	01:41:46
> Run-Time	00:21
> Total	
> Pass	
> Fail	
> Valgrind-Errors	
> Valgrind-Loss	
> Details	vncregress-2022-03-29-01:41:25.txt
> Log	make-2022-03-29-01:41:25.out.bz2
> Memory	
> For details, please contact louberger

@gpnaveen Can you please take a look at it and work with @louberger if needed to understand and fix it?

choppsv1 reviewed (changes requested):
The huge number of test changes are unneeded and should be removed. `mgmtd` is taking over for `staticd` so it should be auto-loaded just as `staticd` is currently in `lib/topotest.py`. Additionally the `staticd`'s use of `backup_config` as zebra.conf should be moved from the `staticd` component to the `mgmtd` component. With these 2 changes *all* of these test changes (what ~100 files?) can be removed.

idryzhov commented on staticd/static_nb.c:
```diff
@@ -60,7 +60,9 @@ const struct frr_yang_module_info frr_staticd_info = {
 		{
 			.xpath = "/frr-routing:routing/control-plane-protocols/control-plane-protocol/frr-staticd:staticd/route-list/path-list/frr-nexthops/nexthop",
 			.cbs = {
+#ifndef INCLUDE_MGMTD_VALIDATE_ONLY
```
Is there any reason for linking `static_nb.c` into mgmtd other than doing mgmtd-side validation?

pushpasis replied:
I think if we remove the need for local validations on MGMTd this file doesn't need to be linked with MGMTd anymore. Will try it out. But first we need consesnsus on wether to keep local validations and remove remote validations, or vice versa. Hoping we can hash this out first in the upcoming meeting I requested.

choppsv1 replied:
It uses it to determine what the component is "subscribed" to I believe. I mentioned in slack that I think it would be passable to move the ifdefs to this file to remove the callback references (so that code doesn't need to be linked in) basically b/c this file is not pretty to start and doesn't really contain functional code (it's just a generated structure) so they are a lot less disruptive.

choppsv1 replied:
FWIW the git patch I submitted to fix remote validation didn't remove local validation from mgmtd, it was very little code to leave it in there. My aim here isn't to remove local validation, but rather to not use it by default and have the initial review and porting effort use remote validation b/c it minimized the component porting changes and review burden (for this PR as well as future porting PRs)

pushpasis commented on mgmtd/mgmt_db.c:
```diff
@@ -0,0 +1,647 @@
+const char *mgmt_db_names[MGMTD_DB_MAX_ID + 1] = {
+	MGMTD_DB_NAME_NONE,	/* MGMTD_DB_NONE */
+	MGMTD_DB_NAME_RUNNING,     /* MGMTD_DB_RUNNING */
+	MGMTD_DB_NAME_CANDIDATE,   /* MGMTD_DB_RUNNING */
```
ACK. Will change it.

pushpasis commented on mgmtd/mgmt_db.c:
```diff
@@ -0,0 +1,647 @@
+
+#ifdef REDIRECT_DEBUG_TO_STDERR
+#define MGMTD_DB_DBG(fmt, ...)                                                 \
+	fprintf(stderr, "%s: " fmt "\n", __func__, ##__VA_ARGS__)
```
This was done for testing at the time of initial development, to directly print the logs directly on to the console. Especially at the time of daemon bringup without needing to call zlog_debug. This is very useful for debugging daemon bringup issues. zlog_debug needs the logging configurations to be applied on the daemon first. By then daemon initializations will be over.

pushpasis replied:
Anyhow this is disabled by default. One will have to define REDIRECT_DEBUG_TO_STDERR, in some header file to to make this effective.

pushpasis commented on lib/command.h:
```diff
@@ -448,6 +448,11 @@ struct cmd_node {
+#define MGMTD_FRNTND_ADPTR_STR "MGMTD Frontend Adapter information\n"
+#define MGMTD_TRXN_STR "MGMTD Transaction information\n"
+#define MGMTD_DB_STR "MGMTD Database information\n"
 
```
ACK. Will take care of it.

pushpasis commented on debian/control:
```diff
@@ -48,6 +48,7 @@ Recommends: frr-pythontools
 Suggests: frr-doc
 Conflicts: pimd,
            quagga,
+           quagga-mgmtd,
```
I have asked @gpnaveen to understand the changes you requested. Will ask him again to reach out to you for further clatrifications needed.

pushpasis commented on mgmtd/mgmt_frntnd_adapter.h:
```diff
@@ -0,0 +1,200 @@
+ */
+
+#ifndef _FRR_MGMTD_FRNTND_ADAPTER_H_
+#define _FRR_MGMTD_FRNTND_ADAPTER_H_
```
The challenge will be now to modify the original commits. Is it okay to do this change in a 9th commit. These abbreviations are spread out across 5 different commis. It will be hard to modify the individual commits using 'git rebase -I'. But definitely can do the abbreviations replacement in a separate commit. Will that be okay with you? Hope anyone else will not have a problem too.

choppsv1 replied:
BTW before you do this, what are we going with? Might as well argue about the color of the bikeshed during the local/remote valiadtion meeting too.. I'm with @idryzhov I'd like to see "fe" and "be" used. There's even precedent.. If you google "frontend backend abbreviation" the one that pops up is "febe". FWIW it's a good idea to do a google search anytime you pick an abbreviation, that way you are more likely to not be dreaming up your own when a standard exists (e.g., transaction => txn came up for me)..

pushpasis commented on mgmtd/mgmt_frntnd_adapter.h:
```diff
@@ -0,0 +1,200 @@
+	struct mgmt_commit_stats cmt_stats;
+	struct mgmt_setcfg_stats setcfg_stats;
+
+	struct mgmt_frntnd_adptr_list_item list_linkage;
```
Got it. Same question as above. Can I do the replacement in a separate commit?

pushpasis commented on mgmtd/mgmt_frntnd_server.c:
```diff
@@ -0,0 +1,174 @@
+
+		if (mgmt_frntnd_listen_fd) {
+			close(mgmt_frntnd_listen_fd);
+			mgmt_frntnd_listen_fd = 0;
```
Accepted. Will take care of it.

pushpasis commented on mgmtd/mgmt_frntnd_server.c:
```diff
@@ -0,0 +1,174 @@
+	union sockunion su;
+
+	if (!mgmt_frntnd_listen_fd)
+		return -1;
```
Sure. Will take care of it.

pushpasis commented on mgmtd/mgmt_frntnd_server.c:
```diff
@@ -0,0 +1,174 @@
+	return;
+
+mgmt_frntnd_server_start_failed:
+	if (sock)
```
ACK. Will take care of it.

pushpasis commented on mgmtd/mgmt_frntnd_server.c:
```diff
@@ -0,0 +1,174 @@
+		close(sock);
+
+	mgmt_frntnd_listen_fd = 0;
+	exit(-1);
```
ACK. Will take care of it.

eqvinox commented on doc/user/mgmtd.rst:
```diff
@@ -0,0 +1,376 @@
+  - This is a copy of Running database that is stored in persistent
+    storage and is used to load configurations on Running database during
+    MGMT daemon startup.
+  - Data cannot be edited/retrieved directly via Frontend interface.
```
There needs to be a little more documentation here, mostly regarding `frr_startup.json`.  (It's mentioned further down, but not sufficient.)

Could you please explain:
- when exactly is `frr_startup.json` written. Is it only `mgmt commit-apply`?
- when is it read? Only at startup?
- is it necessary for the user to back up this file? What happens when the file is lost or corrupted?
- what happens when the user deletes all config in `/etc/frr` but leaves/forgets `frr_startup.json` - will config come back even though it was removed from `/etc/frr`? Can there be a conflict?
- how does `frr version X.Y` & `frr profile <datacenter|traditional>` work?

These docs should be added ASAP, ideally before merge, if not then shortly after.

eqvinox commented:
Superseded, closing.
